server/server.go

```diff
@@ -290,9 +452,100 @@ func (cs *channelStore) count() int {
 }
 
 type channel struct {
-	name  string
-	store *stores.Channel
-	ss    *subStore
+	nextSequence uint64
+	leader       uint32
+	name         string
+	store        *stores.Channel
+	ss           *subStore
+	raft         *raft.Raft
+	raftStore    *raftboltdb.BoltStore
+	lTimestamp   int64
+}
+
+// Apply log is invoked once a log entry is committed.
+// It returns a value which will be made available in the
+// ApplyFuture returned by Raft.Apply method if that
+// method was called on the same Raft node as the FSM.
+func (c *channel) Apply(l *raft.Log) interface{} {
+	op := &spb.RaftOperation{}
+	if err := op.Unmarshal(l.Data); err != nil {
+		panic(err)
+	}
+	switch op.OpType {
+	case spb.RaftOperation_Publish:
+		for _, msg := range op.PublishBatch.Messages {
+			if _, err := c.store.Msgs.Store(msg); err != nil {
+				panic(fmt.Errorf("failed to store replicated message %d on channel %s: %v",
+					msg.Sequence, c.name, err))
+			}
+		}
+	default:
+		panic(fmt.Sprintf("unknown op type %s", op.OpType))
+	}
+	return nil
+}
+
+// pubMsgToMsgProto converts a PubMsg to a MsgProto and assigns a timestamp
+// which is monotonic with respect to the channel.
+func (c *channel) pubMsgToMsgProto(pm *pb.PubMsg, seq uint64) *pb.MsgProto {
+	m := &pb.MsgProto{
+		Sequence:  seq,
+		Subject:   pm.Subject,
+		Reply:     pm.Reply,
+		Data:      pm.Data,
+		Timestamp: time.Now().UnixNano(),
+	}
+	if c.lTimestamp > 0 && m.Timestamp < c.lTimestamp {
+		m.Timestamp = c.lTimestamp
+	}
+	c.lTimestamp = m.Timestamp
+	return m
+}
+
+// isLeader indicates if this node is the channel leader when in clustered mode.
+func (c *channel) isLeader() bool {
+	return atomic.LoadUint32(&c.leader) == 1
+}
+
+// Snapshot is used to support log compaction. This call should
+// return an FSMSnapshot which can be used to save a point-in-time
+// snapshot of the FSM. Apply and Snapshot are not called in multiple
+// threads, but Apply will be called concurrently with Persist. This means
+// the FSM should be implemented in a fashion that allows for concurrent
+// updates while a snapshot is happening.
+func (c *channel) Snapshot() (raft.FSMSnapshot, error) {
```

Comment:
I do not see this function and Restore() being called. I assume this is invoked from raft. From code coverage it looks like these are not covered. Would it be possible to have a test that shows the effect of snapshot and restore?

Reply:
Yes, this is called by Raft. We can probably (and should) test this. Will address...